In this file I'm going to use my previously made code to simulate a poker play

In [ ]:
import card_deck as cd
import cards as card
from fractions import Fraction
import sympy as sp

: 

In [ ]:
simple_deck = cd.Deck(1)
simple_deck.shuffle()
for cards in simple_deck.cards: 
    print(cards.rank)

Ace
Ace
Queen
4
2
Ace
7
9
4
10
8
King
Jack
4
9
9
Jack
6
King
King
Queen
6
Jack
7
6
Queen
3
10
6
4
Jack
2
Queen
9
2
King
2
5
5
3
3
8
10
8
7
8
5
3
5
10
Ace
7


In [ ]:
simple_deck = cd.Deck(1)
for x in(("6","Diamonds"),("4","Hearts")):
    simple_deck.remove_card(x[0], x[1])
for cards in simple_deck.cards: 
    print(cards.display())

2 of Hearts
3 of Hearts
5 of Hearts
6 of Hearts
7 of Hearts
8 of Hearts
9 of Hearts
10 of Hearts
Jack of Hearts
Queen of Hearts
King of Hearts
Ace of Hearts
2 of Diamonds
3 of Diamonds
4 of Diamonds
5 of Diamonds
7 of Diamonds
8 of Diamonds
9 of Diamonds
10 of Diamonds
Jack of Diamonds
Queen of Diamonds
King of Diamonds
Ace of Diamonds
2 of Clubs
3 of Clubs
4 of Clubs
5 of Clubs
6 of Clubs
7 of Clubs
8 of Clubs
9 of Clubs
10 of Clubs
Jack of Clubs
Queen of Clubs
King of Clubs
Ace of Clubs
2 of Spades
3 of Spades
4 of Spades
5 of Spades
6 of Spades
7 of Spades
8 of Spades
9 of Spades
10 of Spades
Jack of Spades
Queen of Spades
King of Spades
Ace of Spades


PROBABILITY OF HAVING A COUPLE WITH TWO DIFFERENT VALUES IN THE HAND MADE WITH X DECKS:


In [ ]:
def is_couple( number_one, number_two, card_in_pool,nr_decks=2): #nr_decks,number_one,number_two, card_in_pool
    """I'm checking if the numbers 3 and 2 are in the next 5 cards, if  one condition is True then it returns 1"""
    my_deck = cd.Deck(nr_decks)
    hand = (str(number_one), str(number_two))

    my_deck.shuffle()
    for x in range(card_in_pool):
        my_deck.deal_card()

    for cards in my_deck.pool:
        for x in hand:
            if cards.rank == x:
                return 1
    return(0)

def check_couple(number_one, number_two, card_in_pool):
    good_tries = 0
    tries = 0
    for x in range(10000):
        good_tries = good_tries + is_couple(number_one, number_two, card_in_pool)
        tries = tries + 1
    print(good_tries/tries)
    

In [ ]:
check_couple(2,3,5)

0.5752


We can clearly see that with a pool of 10'000 tries the probability converges to 0.57...
so I can clearly say that the probabilty of having at least a couple is 0.57.

But what is the probability of having a couple in the first round? And if I hadn't got a couple in the first round, what would be the probability?

In [ ]:
"""functions"""

'functions'

In [ ]:
"""tries"""
#couple served in the first round:
check_couple(2,3,3)

#couple served in the fourth or fifth round 
check_couple(2,3,2)

#get a couple in the fifth:
check_couple(2,3,1)


0.404
0.287
0.1557


"""results"""
I can clearly see that the probabilty of having a couple in a two deck situationship (without having a couple at the beginning) is:
0.57
then:
0.404;
0.287;
0.1557


But what is the probability of having a couple before playing?

In [ ]:
"""functions"""
def couple_at_beginning():
    my_deck = cd.Deck(2)
    my_deck.shuffle()
    for x in range(2):
        my_deck.deal_card()
    if my_deck.pool[0].rank == my_deck.pool[1].rank:
        return(1)
    else:
        return(0)
    
    
def is_couple_at_beginning():
    a=0
    b=0
    for x in range(10000):
        a= a+couple_at_beginning()
        b= b+1
    print(a/b)

In [ ]:
"""tries"""

is_couple_at_beginning()

0.0695


"""results"""
We can see that the probability is near 0.069-0.072
(very low)

Then what is the absolute possibility of having a couple?



In [ ]:
"""functions"""
def couple_absolute(nr_card_dealt):
    my_deck = cd.Deck(2)
    my_deck.shuffle()
    for x in range(nr_card_dealt):
        my_deck.deal_card()
    a = -1
    for x in my_deck.pool:
        my_list = my_deck.pool.copy()
        my_list.remove(x)
        for y in range(len(my_list)):
            if x.rank == my_list[y].rank :
                return(1)
    return(0)

def couple_absolute_possibility(nr_card_dealt):
    a=0
    b=0
    for x in range(10000):
        a= a+couple_absolute(nr_card_dealt)
        b= b+1
    print(str(a)+" over " +str(b))
    print(a/b)
    



In [ ]:
"""tries"""
couple_absolute_possibility(7)
couple_absolute_possibility(9)
couple_absolute_possibility(11)
couple_absolute_possibility(13)

8324 over 10000
0.8324
9654 over 10000
0.9654
9971 over 10000
0.9971
9998 over 10000
0.9998


"""results"""
We can clearly see that the possibility that there is at least a couple in the groud if I deal 7 card is:
0.75-80
if 8:
0.953;
0.9958
0.9996

But the results in my opinion are suspect because I clearly remember times in which I had 4 players and on the table there wasn't a couple
I could use mathematics to prove it is wrong or right:

probability of NOT having a couple on the floor:
$$
\text{n = card drawn }
\{\forall n \in \mathbb{N} | x<14\}
\frac{(n-1)!}{(n)^{(n-1)}}
$$

so, for x = 13 then the probability is $\frac{1}{20000}$ so, I have confutated myself


"""Breaking point"""


I have only calculated the possibility of having a couple but I can surely adapt the function to other situationship:
(double couple, tris, flush, straight, full house, poker, straight flush,royal flush)


But before I do some claculation I must ask myself what could be useful to do.
This question has multiple answer:
What is my probability of having some good combination
which probability has my opponent of having a good combination

and the synthesys of it all is:
    therefore: how much money am I disposed to bet ort, how much money am I disposed to risk in it?
    
Before I start to navigate in this sea I would say that the synthesys has not a good and clear solution. Because the odds could be all against me and therefore I could always lose. But if I can divide the odds against me and therefore know which kind of bet could in the long term make me have a stable win then I will be more likely to win.


In [29]:
import game_assembler as ga
import card_deck as cd
import cards
from hand import Hand
game = ga.GameAssembler(2,cd.Deck(2))
game.deal_card_to_player()
print(game.players_hands)
game.deal_card_to_player()
print(game.players_hands[0].show_card(0))

[<hand.Hand object at 0x7066cc2671a0>, <hand.Hand object at 0x7066cc0a1070>]


AttributeError: 'Hand' object has no attribute 'show_card'

In [36]:
import game_assembler as ga
import card_deck as cd
import cards
from hand import Hand

game_assembler = ga.GameAssembler(2, cd.Deck(2))
player_hand = game_assembler.players_hands[0]
print(player_hand[0])          # prints e.g. "Ace of Spades"
print(player_hand.show_card(0))# same, returns string

TypeError: 'Hand' object is not subscriptable